### DISCUSSION TITLE
Cleaning the mesh fragments in holes (clean cutting)

### DISCUSSION THREAD

**Discussion by drathin4** [libigl/libigl#1921] [Q&A]
I have written code for MeshBoolean which does a series of unite/subtract operations. There are holes present in the output mesh which are typically cut cleanly, but occassionally have some leftover material. This seems to be due to incorrectly positioned input meshes during Boolean subtract operations. Is it possible to trim/clean the mesh fragments in the holes as a post-processing/repair step ?

Attached herewith are the images for illustration.

![image](https://user-images.githubusercontent.com/79855313/142422390-d0347f93-81db-4edb-a46b-08dc4e813d74.png)

![image](https://user-images.githubusercontent.com/79855313/142422547-faa441c7-5697-4d75-9824-a758d11756c3.png)

![image](https://user-images.githubusercontent.com/79855313/142422708-dfdf8c2e-1b28-48d7-990d-3ddb1b740be6.png)

**Comment by alecjacobson** [comment]
To understand correctly,  these images come from mal-formed _inputs_ to
mesh Boolean operations?

**Reply by drathin4** [reply]
They're not exactly malformed inputs, rather incorrectly positioned. And only some of the inputs are incorrectly positioned. Typically, the holes are cut cleanly barring a few. The below image demonstrates that.

The positions of the inputs are very difficult to correct in the upstream code and hence I'm thinking of implementing a post-processing step that will cut the holes cleanly (clear the fragments present in the holes).

![image](https://user-images.githubusercontent.com/79855313/142460365-2d164a0b-1710-48d2-a756-00fc9b5873a0.png)

**Comment by alecjacobson** [comment]
The "morphological closing" is probably the operation that you would need
to do here. It's easiest implemented with voxels or distance functions
(either resulting in a complete remeshing of the surface). libigl doesn't
support these directly but you might follow something like
https://github.com/alecjacobson/gptoolbox/blob/master/mesh/closing.m
